packages/jupyter-ai/jupyter_ai/personas/directories.py

```diff
@@ -0,0 +1,98 @@
+from pathlib import Path
+from typing import Optional
+
+
+def find_dot_dir(
+    dir: str, dot_dir: str, root_dir: Optional[str] = None
+) -> Optional[str]:
+    """
+    Find the nearest dot directory by traversing up from the given directory.
+
+    Args:
+        dir (str): The starting directory path
+        dot_dir (str): The dot directory name to search for (e.g., '.jupyter', '.git')
+        root_dir (Optional[str]): The root directory to stop searching at.
```

Comment:
[nit] I would recommend renaming `root_dir` to `contents_root` throughout this branch since this names collides with the existing concept of a root directory, referring to `/` or `C:\\`.